### PR TITLE
FIX #23 - Fixing TypeError exception

### LINKED ISSUE #23: TypeError when deploying
I've tried this on microk8s with juju 2.9-rc6 and microk8s v1.20.2,. and I got a different failure:

```
application-graylog: 13:22:07 INFO juju.worker.caasoperator.uniter.graylog/0 awaiting error resolution for "relation-changed" hook
application-graylog: 13:22:17 INFO juju.worker.caasoperator.uniter.graylog/0 awaiting error resolution for "relation-changed" hook
application-graylog: 13:22:18 ERROR unit.graylog/0.juju-log mongodb:3: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 223, in <module>
    main(GraylogCharm)
  File "/var/lib/juju/agents/unit-graylog-0/charm/venv/ops/main.py", line 401, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-graylog-0/charm/venv/ops/main.py", line 140, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-graylog-0/charm/venv/ops/framework.py", line 234, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-graylog-0/charm/venv/ops/framework.py", line 678, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-graylog-0/charm/venv/ops/framework.py", line 723, in _reemit
    custom_handler(event)
  File "./src/charm.py", line 127, in _on_mongodb_relation_changed
    self._configure_pod()
  File "./src/charm.py", line 215, in _configure_pod
    spec = self._build_pod_spec()
  File "./src/charm.py", line 173, in _build_pod_spec
    'GRAYLOG_HTTP_PUBLISH_URI': self.external_uri,
  File "./src/charm.py", line 56, in external_uri
    ingress = str(self.model.get_binding('graylog').network.ingress_address)
  File "/var/lib/juju/agents/unit-graylog-0/charm/venv/ops/model.py", line 506, in network
    self._network = Network(self._backend.network_get(self.name, self._relation_id))
  File "/var/lib/juju/agents/unit-graylog-0/charm/venv/ops/model.py", line 545, in __init__
    for address_info in interface_info.get('addresses', []):
TypeError: 'NoneType' object is not iterable
application-graylog: 13:22:18 ERROR juju.worker.caasoperator.uniter.graylog/0.operation hook "mongodb-relation-changed" (via hook dispatching script: dispatch) failed: exit status 1
```

This left MongoDB in `hook failed: "mongodb-relation-changed"`.

The steps I took were (I used charms from the charm store/charmhub where possible):
```
juju add-model graylog micro
juju deploy cs:~charmcraft/graylog
cd ~/repos/
mkdir mongodb-operator
cd mongodb-operator/mongodb-operator/
git pull
charmcraft build
juju deploy ./mongodb.charm
juju deploy ./mongodb.charm --resource mongodb-image=mongo:4.4.1
juju deploy cs:~charmcraft/elasticsearch
juju add-unit -n 2 mongodb
# Wait for juju status to settle, mongodb and elasticsearch active and idle
# graylog blocked with message "Need mongodb and Elasticsearch relations."
juju add-relation graylog mongodb
juju add-relation graylog elasticsearch
```

_Originally posted by @mthaddon in https://github.com/canonical/graylog-operator/issues/21#issuecomment-791388989_

### PR DESCRIPTION
**Pull request by Abuelodelanada**

### CODE REVIEW

**Comment by jlounder** [comment]
I would prefer an alternate approach. 

- external_uri can raise its own exception when it encounters an error fetching the type error (or return false/None)
- configure_pod can catch the exception (IngressUnavailable, or something like that) and then return false - indicating it failed to configure the pod
- each event handler calling configure pod can call event.defer, if it needs to. 

This should allows a lot more flexibility at each layer, if any of these functions need to be used in other parts of the code.

File: src/charm.py [hunk 1]
```diff
@@ -54,12 +54,15 @@ def bind_address(self):
...
-        ingress = str(self.model.get_binding('graylog').network.ingress_address)
-        port = self.model.config['port']
-        return 'http://{}:{}/'.format(ingress, port)
+        try:
+            ingress = str(self.model.get_binding('graylog').network.ingress_address)
+            port = self.model.config['port']
+            return 'http://{}:{}/'.format(ingress, port)
+        except TypeError:
+            event.defer()
+            return "http://"
```
Comment (balbirthomas) [inline src/charm.py:65]: It is not clear why you need a return if you are deferring the event.

**Review by balbirthomas — APPROVED** [review]

**Review by dstathis — APPROVED** [review]

File: src/charm.py (hunk 1 above)
Comment (dstathis) [reply src/charm.py:65]: defer does not end execution

Comment (balbirthomas) [reply src/charm.py:65]: > defer does not end execution

Yes I am aware. The `return "http://"` happens to be the last line of the function though.

**Review by balbirthomas — APPROVED** [review]
May be worth asking in Juju channel why and PEER ingress address would not be available at all times.

File: src/custom_exceptions.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env python3
+# Copyright 2020 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""
+This module has custom exceptions for garylog-operator
```
Comment (dstathis) [inline src/custom_exceptions.py:6]: Typo here "garylog"

**Review by dstathis — CHANGES_REQUESTED** [review]

**Review by dstathis — APPROVED** [review]

**Pull request merged**